### ISSUE TITLE
MedNeXt (versions 1 and 2) experiments, VRAM and runtime reporting

### ISSUE THREAD

**Issue by Shrajan** [MIC-DKFZ/nnUNet#2996]
Hi nnUNet team,

I have some questions regarding the experimental setup and reported results in the `MedNeXt-v1` [1], `MedNeXt-v2` [2] and `nnU-Net Revisited`[3] papers. I have been trying to carefully reproduce the experiments using the official nnU-Net framework, MedNeXt repository, and settings.

1) I would like to ask about GPU VRAM usage reported in Table 1 of the `nnU-Net Revisited` paper [3]. I reproduced the experiments using the official code, same presets and patch size, checkpointing and AMP enabled, and on an A100 40 GB GPU. The runtime scaling between `ResEnc L` and `MedNeXt L k3` is very similar to what is reported in the paper, so I believe my setup is aligned with the benchmark. However, the allocated memory for `MedNeXt L k3` is noticeably higher in my experiments, and in my case MedNeXt uses more memory than `ResEnc L`, while in Table 1 [3] it is reported to use less. `MedNeXt L k5` also utilises similar GPU VRAM as `MedNeXt L k3`, but of course, takes longer to run. 

For ResEnc UNet L on A100 40 GB, my results match closely:
```
Training time ≈ 28 hours (paper: 35 hours)
Allocated CUDA memory ≈ 21.5 GB (paper: 22.7 GB)
Reserved CUDA memory ≈ 22.6 GB
Nvidia-smi ≈ 23.5 GB
```
However, for MedNeXt L k3 on A100 40 GB, I observe:
```
Training time ≈ 60 hours (paper: 68 hours)
Allocated CUDA memory ≈ 24.9 GB (paper: 17.3 GB)
Reserved CUDA memory ≈ 33.3 GB
Nvidia-smi ≈ 34.6 GB
```

I even trianed the MedNeXt models with a RTX 5090 (32 GB) GPU, and even then the VRAM usage did not go below 20 GB.

2) Could you please clarify:
a) Whether any additional settings were used for MedNeXt [3] in the benchmark?
b) How exactly VRAM was measured for Table 1 (allocated vs reserved memory or Nvidia-smi)[3]?
c) Whether a slightly different MedNeXt configuration or commit version was used internally?
d) I also wanted to confirm whether the reported runtime in Table 1 [3] is measured per fold (average), right? Not aggregated over all 5 cross-validation folds.

3) Finally, I noticed that the DSC values of nnUNet, ResEnc UNet L and MedNeXt-v1 trained from scratch differ slightly between the MedNeXt-v1 [1] and MedNeXt-v2 [2] papers (for example on AMOS, KiTS and ACDC). I understand that different splits or protocols may have been used, but I would appreciate clarification on this as well.

Thank you very much for your time and help.

References:
[1] MedNeXt v1 - https://conferences.miccai.org/2023/papers/410-Paper1656.html
[2] MedNeXt v2 - https://arxiv.org/pdf/2512.17774
[3] nnU-Net Revisited - https://papers.miccai.org/miccai-2024/paper/2847_paper.pdf

**Comment by saikat-roy** [comment]
Hi @Shrajan, sorry for the late reply. Can I ask you what your patch size was for the VRAM calculations in your work? You mentioned it was the same - was it specifically `128x128x128`?

Also, one thing I have seen in MedNeXt that might not be an issue in ResEncL is fragmentation. Specifically, I had started using `PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True` to overcome some of the issues with PyTorch reserved memory. Could you do this and see if there's any effect to a) your VRAM usage and b) You inability to run on 40 GB A100s.

> a) Whether any additional settings were used for MedNeXt [3] in the benchmark? 
> b) How exactly VRAM was measured for Table 1 (allocated vs reserved memory or Nvidia-smi)[3]?
> c) Whether a slightly different MedNeXt configuration or commit version was used internally?
> d) I also wanted to confirm whether the reported runtime in Table 1 [3] is measured per fold (average), right? Not aggregated over all 5 cross-validation folds.

I think regarding the benchmarking in nnUNet Revisited itself, @constantinulrich is the best person to answer the settings when computing memory usage. I know that project used the original MedNeXt repository (based on nnUNet-v1) and not for example, porting the MedNeXt architecture to the newer nnUNet-v2. But that in principle should not affect the computation of VRAM usage during training, particularly via `nvidia-smi`.

Additionally, the results in the MedNeXt-v2 paper for your mentioned datasets are fold 0 results for a smaller ablation experiment, while those in the nnUNet Revisited are 5 fold CV results. That is why they differ. I also used a older version of AMOS with 200 samples, so that yet again is another point of difference.

**Comment by saikat-roy** [comment]
Ah yes you're right, we do mention the 200 samples in the dataset size of AMOS in the MedNeXt-v2 paper, but we indeed don't specifically highlight why it is lower. The paper is also a preprint currently and will probably go through a few more iterations in the future.

**Comment by FabianIsensee** [comment]
Have all questions in this issue been resolved? If not, what is missing?

**Comment by Shrajan** [comment]
Hi @FabianIsensee,

Saikat mentioned that `constantinulrich` knows more about the benchmarking in the `nnUNet revisited` paper, and so I am still waiting for answers for these two:
2 (a) Whether any additional settings were used for MedNeXt [3] in the benchmark?
2 (b) How exactly VRAM was measured for Table 1 (allocated vs reserved memory or Nvidia-smi) [3]?

**Comment by FabianIsensee** [comment]
I can answer the VRAM thing. That was simply done with nvidia-smi. For the rest I refer you to @constantinulrich

**Comment by constantinulrich** [comment]
> Whether any additional settings were used for MedNeXt [3] in the benchmark?

No, actually, I used the default repository as it was published based on nnUNet V1. 

Best
Constantin

**Comment by Shrajan** [comment]
Thank you @FabianIsensee, @saikat-roy and @constantinulrich for all your detailed responses.

As a final note, compared to my experiments with the A100 40 GB GPU using the MedNeXt repository, the VRAM memory consumption of `MedNeXt L` models mentioned in the `nnUNet revisited paper` is significantly lower (nearly half) even with checkpointing. The details are listed again below. 

MedNeXt L k3 on A100 40 GB:
```
Training time ≈ 60 hours (paper: 68 hours)
Allocated CUDA memory ≈ 24.9 GB 
Reserved CUDA memory ≈ 33.3 GB
Nvidia-smi ≈ 34.6 GB (paper: 17.3 GB)
```

**Comment by FabianIsensee** [comment]
IIRC MedNext uses gradient checkpointing. Did either of you change the defaults when running your experiments? @constantinulrich do you happen to have the v1 env still somewhere and could double check that 17.3G number?
